File: Leetcode100/leetcode100_python/Searcha2DMatrixII3.py
class Solution:
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        def searchMatrix(up, down, left, right):
            if up > down or left > right:
                return False

            i, j = up, down
            while i <= j:
                m = i + (j - i) // 2
                if matrix[m][left] == target:
                    return True
                elif matrix[m][left] < target:
                    i = m + 1
                else:
                    j = m - 1
            row = j

            i, j = left, right
            while i <= j:
                m = i + (j - i) // 2
                if matrix[row][m] == target:
                    return True
                elif matrix[row][m] < target:
                    i = m + 1
                else:
                    j = m - 1
            col = j

            return searchMatrix(up, row, col + 1, right)
        if not matrix:
            return False
        return searchMatrix(0, len(matrix) - 1, 0, len(matrix[0]) - 1)

if __name__ == "__main__":
    n = int(input())
    matrix = []
    for i in range(n):
        matrix.append([int(num) for num in input().split()])
    target = int(input())
    solution = Solution()
    print(matrix)
    print(solution.searchMatrix(matrix, target))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/search_a_2D_matrix_II/Searcha2DMatrixII3.java
package main.java.com.jiang.leetcode100.search_a_2D_matrix_II;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution3 {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length;
        if(m <= 0) return false;
        int n = matrix[0].length;
        
        int row = m - 1, col = 0;
        while(row >= 0 && col < n) {
            if(matrix[row][col] > target) row--;
            else if(matrix[row][col] < target) col++;
            else return true;
        }
        return false;
    }
}

public class Searcha2DMatrixII3 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        List<String> inputs = new ArrayList<>();
        while(!scanner.hasNext("#")) {
            inputs.add(scanner.nextLine());
        }
        scanner.nextLine();
        int target = scanner.nextInt();
        scanner.close();

        int[][] matrix = new int[inputs.size()][inputs.get(0).split(" ").length];
        for(int i = 0; i < inputs.size(); i++) {
            String[] nums_str = inputs.get(i).split(" ");
            int[] nums = new int[nums_str.length];
            for(int j = 0; j < nums_str.length; j++) nums[j] = Integer.parseInt(nums_str[j]);
            matrix[i] = nums;
        }

        if(new Solution3().searchMatrix(matrix, target)) System.out.println("true");
        else System.out.println("false");
    }
}